File: Leetcode/1. Two Sum/1_Two_Sum.java
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Create a HashMap to store the elements from the 'nums' array as keys
        // and their corresponding indices as values.
        HashMap<Integer, Integer> maps = new HashMap<>();

        // Iterate through each element in the 'nums' array.
        for (int i = 0; i < nums.length; i++) {
            // Calculate the difference between the 'target' value and the current element.
            int diff = target - nums[i];

            // Check if the 'diff' value is already present in the 'maps' HashMap.
            if (maps.containsKey(diff)) {
                // If 'diff' is found, return an array containing the indices of the two
                // elements.
                // 'maps.get(diff)' retrieves the index of the element that complements 'diff',
                // and 'i' is the index of the current element.
                return new int[] { maps.get(diff), i };
            }

            // If 'diff' is not found, add the current element and its index to the 'maps'
            // HashMap.
            maps.put(nums[i], i);
        }

        // If no pair of elements that add up to the 'target' is found, return an empty
        // integer array.
        return new int[] {};

    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = { 2, 7, 11, 15 };
        int target = 9;
        int[] result = solution.twoSum(nums, target);
        System.out.println("[" + result[0] + "," + result[1] + "]");
    }
}

File: Leetcode/1. Two Sum/1_Two_Sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Create an empty dictionary 'maps' to store elements from the 'nums' list as keys
        # and their corresponding indices as values.
        maps = {}

        # Iterate through each element and its index in the 'nums' list.
        for index, num in enumerate(nums):
            # Calculate the difference between the 'target' value and the current element.
            diff = target - num

            # Check if the 'diff' value is already present in the 'maps' dictionary.
            if diff in maps:
                # If 'diff' is found, return a list containing the indices of the two elements.
                # 'maps[diff]' retrieves the index of the element that complements 'diff',
                # and 'index' is the index of the current element.
                return [maps[diff], index]

            # If 'diff' is not found, add the current element and its index to the 'maps' dictionary.
            maps[num] = index

        # If no pair of elements that add up to the 'target' is found, return an empty list.
        return []


class Main:
    def main(self):
        result = Solution().twoSum([2, 7, 11, 15], 9)
        print(result)


if __name__ == "__main__":
    Main().main()
